pkgs/tools/security/pinentry/default.nix

```
@@ -1,103 +1,90 @@
-      '' + optionalString (f == "qt") ''
-        wrapQtApp ${placeholder f}/bin/${binary}
-      '')) + ''
-      ln -sf ${placeholder (head enabledFlavors)}/bin/pinentry-${flavorInfo.${head enabledFlavors}.bin} $out/bin/pinentry
-    '';
-
-  outputs = [ "out" ] ++ enabledFlavors;
-
-  passthru = { flavors = enabledFlavors; };
+        info = flavorInfo.${f};
+        flag = flavorInfo.${f}.flag or null;
+      in
+        lib.optionalString (flag != null)
+          (mkEnable (lib.elem f buildFlavors) ("pinentry-" + flag));
+
+    in stdenv.mkDerivation rec {
+      pname = "pinentry-${pnameExtra}";
+      version = "1.1.1";
+
+      src = fetchurl {
+        url = "mirror://gnupg/pinentry/pinentry-${version}.tar.bz2";
+        sha256 = "0zx5vg6wws2sp2yxwi01b8i1pnsqkydncpj7x0p8xl9y05ja04nd";
+      };
+
+      nativeBuildInputs = [ pkg-config autoreconfHook ]
+        ++ lib.concatMap (f: flavorInfo.${f}.nativeBuildInputs or []) buildFlavors;
+      buildInputs = [ libgpgerror libassuan libcap libsecret ]
+        ++ lib.concatMap (f: flavorInfo.${f}.buildInputs or []) buildFlavors;
+
+      dontWrapGApps = true;
+      dontWrapQtApps = true;
+
+      patches = [
+        ./autoconf-ar.patch
+      ] ++ lib.optionals (lib.elem "gtk2" buildFlavors) [
+        (fetchpatch {
+          url = "https://salsa.debian.org/debian/pinentry/raw/debian/1.1.0-1/debian/patches/0007-gtk2-When-X11-input-grabbing-fails-try-again-over-0..patch";
+          sha256 = "15r1axby3fdlzz9wg5zx7miv7gqx2jy4immaw4xmmw5skiifnhfd";
+        })
+      ];
+
+      configureFlags = [
+        (mkWith   (libcap != null)    "libcap")
+        (mkEnable (libsecret != null) "libsecret")
+      ] ++ (map mkFlavorFlag (lib.attrNames flavorInfo));
+
+      postInstall =
+        lib.optionalString (lib.elem "gnome3" buildFlavors) ''
+          wrapGApp $out/bin/pinentry-gnome3
+        '' + lib.optionalString (lib.elem "qt" buildFlavors) ''
+          wrapQtApp $out/bin/pinentry-qt
+        '';
+
+      passthru = { flavors = buildFlavors; };
+
+      meta = with lib; {
+        homepage = "http://gnupg.org/aegypten2/";
+        description = "GnuPG’s interface to passphrase input";
+        license = licenses.gpl2Plus;
+        platforms = platforms.all;
+        longDescription = ''
+          Pinentry provides a console and (optional) GTK and Qt GUIs allowing users
+          to enter a passphrase when `gpg' or `gpg2' is run and needs it.
+        '';
+        maintainers = with maintainers; [ ttuegel fpletz ];
+      };
+    };
+in
 
-  meta = with lib; {
-    homepage = "http://gnupg.org/aegypten2/";
-    description = "GnuPG’s interface to passphrase input";
-    license = licenses.gpl2Plus;
-    platforms = platforms.all;
-    longDescription = ''
-      Pinentry provides a console and (optional) GTK and Qt GUIs allowing users
-      to enter a passphrase when `gpg' or `gpg2' is run and needs it.
-    '';
-    maintainers = with maintainers; [ ttuegel fpletz ];
-  };
+{
```

This is not too override-friendly. Would be nice to at least have
```suggestion
{
  inherit buildPinentry;
```

---

I do understand your concern here. Unfortunately this would break the simple discoverability of available flavors in the NixOS gnupg module because it just uses `lib.attrNames pkgs.pinentryFlavors`. Do you have any suggestions?

---

`buildPinentry` could be refactored into a "regular" package that takes its arguments from `callPackage`.

```diff
-  curses = buildPinentry "curses" [ "curses" "tty" ];
+  curses = callPackage ./pinentry.nix { name = "curses"; flavors = [ "curses" "tty" ]; };
```

This makes the `name` and `flavors` overridable via `pinentry.override`.
I suspect that what's `callPackages` now should become a plain `import`.

---

> `buildPinentry` could be refactored into a "regular" package that takes its arguments from `callPackage`.
> 
> ```diff
> -  curses = buildPinentry "curses" [ "curses" "tty" ];
> +  curses = callPackage ./pinentry.nix { name = "curses"; flavors = [ "curses" "tty" ]; };
> ```
> 
> This makes the `name` and `flavors` overridable via `pinentry.override`.
> I suspect that what's `callPackages` now should become a plain `import`.

I agree with this direction. It should be simpler to just have one layer of callPackages done inside the all-packages.nix:
```nix
pinentry-curses = callPackages ../tools/security/pinentry {
  suffix = "curses"; flavors = [ "curses" "tty" ];
  libcap = if stdenv.isDarwin then null else libcap;
};
pinentry-foo = /* ... */;
pinentry-bar = /* ... */;
```
The `libcap` setting may repeat, and that's fine IMO.

---

`libcap` logic could move into the package file.

I do think we need the singular form.

---

> I do think we need the singular form.

We could also do without it. Btw, it's not comprehensive: you can't build `pinentry_mac` from the same derivation.

---

Just rebased this PR and will work on an override-able solution. Thanks for the pointers.